draft-ietf-quic-v2.md

```diff
@@ -178,23 +178,22 @@ server MUST include the relevant transport parameters to validate that the
 server sent the Retry and the connection IDs used in the exchange, as described
 in {{Section 7.3 of QUIC}}.
 
-The server SHOULD start sending its Initial packets using the negotiated
-version as soon as it decides to change. Before the server is able to process
-transport parameters from the client, it might need to respond to Initial
-packets from the client. For these packets the server uses the original version.
+As the server cannot send CRYPTO frames until it has processed the client's
+transport parameters, it MUST send all CRYPTO frames in Initial packets using
+the negotiated version. The client learns the negotiated version from the first
+Initial CRYPTO frame it successfully processes, and MUST discard any packets
+with a higher packet number that use the original version.
 
-A client learns the negotiated version from the version field of the first
-successfully processed Initial packet that is not equal to the original version.
-If it processes no such packet, it learns the negotiated version from the first
-successfully processed Handshake packet, which might indicate that no version
-change takes place.
+Before the server is able to process transport parameters from the client, it
+might need to respond to Initial packets from the client. For these packets, the
+server uses the original version.
 
 Once the client has learned the negotiated version, it SHOULD send subsequent
```

Comment:
Since we're opening the 2119 box, should we also change this? This adds a requirement for the server to install negotiated version initial keys even though they're not useful in practice

Reply:
I am concerned that the obvious path here is for the client to send INITIAL ACK with the negotiated version. I'll check with implementers, but I really don't want to break existing implementations.

Reply:
The server only has to worry about 2 versions, while the client has to worry about everything in the Supported Versions field.

Reply:
I haven't implemented this myself, but it would be much easier to implement in Google's implementation if the server never had to keep multiple sets of keys per encryption level.

Reply:
Though this is likely to be an unusual case, Initial packets might need to be sent by the client if the server Initial is very large (think PQ key exchange).  If the server has chosen a different version, the client can choose to send them in the original or negotiated version.  Or at least as far as the specification is currently written.

The server probably has to keep the keys for the original version, so David is justified in asking whether we should just stick to that version for Initial packets.  Having the server install new keys for the negotiated version is fairly trivial, but it's not strictly needed.  Encouraging the clean break on the client end means that the server needs to install a second set of keys.

Forcing the client to stick with the original version would be a change for me.  The implementation I wrote has the client switch over to using the negotiated version as described here; the server adds a second set of keys for Initial packets so that it can continue to process both old and new packets.  On the client side keys for all compatible versions are installed from the beginning.  A change here will result in removing some code, though see below.

I don't know how easy it would be to deploy, however.  There are deployed clients that do the switchover and servers that handle it.  A server that was written to this proposed design would not work with those existing clients.

This is only a programming convenience for server implementations, as they are already in the position where - even if they only support one version - they need to generate Initial keys on the fly as packets arrive.  The only real wrinkle here is associating more than one key with connection state.  For servers that share code with clients though, a client implementation needs to support multiple receive keys when it enables compatible version negotiation.

So I'm not seeing a lot to be gained.  Servers probably need to accept Initials with both keys for the foreseeable future due to the deployed base.  Savings, such as they are, are a tiny amount of computation and some engineering work that a lot of code has to take on anyway.

Separately, this creates a special case that seems potentially awkward for generic code.  The client would have to send v1 packets on a connection that it knows is v2.  For v1 and v2, this doesn't make any real difference, but that's not something we can guarantee for any arbitrary pair of versions (yes, I know that this is the v2 spec).  Mapping out compatibility for the first flight is one thing, but that doesn't necessarily extend to other stuff that might be sent in Initial packets (CONNECTION_CLOSE for instance).  Building a clean switchover has a chance of applying to many QUIC versions, but this special case might not.  I'd like to avoid having to think about how to accommodate that in the design.

For me, the special case thing is not that important.  Initial packets aren't useful for much, so I could maybe tolerate code that only ever works for QUIC v1 and v2 in a pinch.  The deployment thing seems far more of a problem though.  Relative to what a server gains for the change, I'd lean heavily toward not changing the rules.

Reply:
That's fair. I agree that it's quite late in the process so I won't push this further.